lean4-infoview/src/infoview/userWidget.tsx

```diff
@@ -0,0 +1,54 @@
+import * as React from 'react';
+
+import { RpcContext } from './contexts';
+import { DocumentPosition, mapRpcError, useAsync } from './util';
+import { ErrorBoundary } from './errors';
+import { Widget_getWidgetSource, UserWidget } from './rpcInterface';
+
+function dynamicallyLoadComponent(hash: string, code: string,) {
+    return React.lazy(async () => {
+        const file = new File([code], `widget_${hash}.js`, { type: 'text/javascript' })
+        const url = URL.createObjectURL(file)
+        return await import(url)
+    })
+}
+
+const componentCache = new Map()
+
+interface UserWidgetProps {
+    pos: DocumentPosition
+    widget: UserWidget
+}
+
+export function UserWidget({ pos, widget }: UserWidgetProps) {
+    const rs = React.useContext(RpcContext);
+    if (!pos) {
+        return <>Waiting for a location.</>
+    }
+    const [status, component, error] = useAsync(
+        async () => {
+            if (componentCache.has(widget.hash)) {
+                return componentCache.get(widget.hash)
```

Comment:
Sorry I don't understand this.

Reply:
When the position changes, `UserWidget` first shows the "pending" status, and starts a promise to fetch the javascript (via `useAsync`).  When this promise is finished, the `UserWidget` is rerendered.  The `UserWidget` function is called very often, and every call with a different position requires a second update.

Most of the time, the widget javascript is already in the cache (because there's like one widget and it's loaded once).  Then we don't need to do a second update.

I'm not sure if there's an easy and nice way to make `useAsync` return the cached value immediately (without going through the Promise.then -> set React reference -> React update dance).

Reply:
ok I think the solution here is remove async function's dependency on pos, it just needs _a_ pos to find a snapshot.

Reply:
fixed